// @edwinspire/express-pgapi/pgListen.js
const dotenv = require("dotenv");
dotenv.config();
const connectionString = process.env.DATABASE_URL;
const EventEmitter = require('events');
const { Client } = require('pg');

export class pgListen extends EventEmitter {
    constructor(listen_notification_list) {
        super();
        if (listen_notification_list && listen_notification_list.length > 0) {
            let events = listen_notification_list.map((event) => {
                return 'LISTEN "' + event + '";';
            });

            this.query = events.join(' ');

            //console.log(query);
            this.client = new Client({
                connectionString: connectionString,
                ssl: {
                    rejectUnauthorized: false
                }
            });

            this.connect();

        }


    }

    connect() {
        this.client.connect();

        this.client
            .query(this.query)
            .then(result => {
                console.log('START LISTEN')
            })
            .catch((e) => {
                console.trace(e.stack);
            })


        this.client.on('error', err => {
            console.trace(err.stack)
        })

        this.client.on('notification', not => {
            this.emit('notification', not);
        })

    }

}
